forge/src/main/java/me/lucko/luckperms/forge/util/BrigadierInjector.java

```diff
@@ -139,7 +149,26 @@ private InjectedPermissionRequirement(String permission, Predicate<CommandSource
         public boolean test(CommandSourceStack source) {
             if (source.getEntity() instanceof ServerPlayer) {
                 ServerPlayer player = (ServerPlayer) source.getEntity();
-
+                // If player is still connecting and has not been added to world then initialise capability
+                UserCapabilityImpl userCapability = UserCapabilityImpl.get(player);
+                if (!player.isAddedToWorld() && !userCapability.initialised()) {
+                    GameProfile profile = player.getGameProfile();
+                    User user = this.plugin.getUserManager().getIfLoaded(profile.getId());
+
+                    if (user == null) {
+                        this.plugin.getLogger().warn("User " + profile.getId() + " - " + profile.getName() +
+                                " doesn't currently have data pre-loaded, but they have been processed before in this session.");
+
+                        Component component = TranslationManager.render(Message.LOADING_STATE_ERROR.build(), player.getLanguage());
+                        if (this.plugin.getConfiguration().get(ConfigKeys.CANCEL_FAILED_LOGINS)) {
+                            player.connection.disconnect(ForgeSenderFactory.toNativeText(component));
+                            return false;
+                        }
+                    }
+                    // initialise capability
+                    userCapability.initialise(user, player, this.plugin.getContextManager());
+                    this.plugin.getContextManager().signalContextUpdate(player);
+                }
```

Comment:
It looks like this has just been copy pasted from

```java
@SubscribeEvent(priority = EventPriority.HIGHEST)
public void onPlayerLoadFromFile(PlayerEvent.LoadFromFile event) {}
```

in ForgeConnectionListener.java.

Maybe we are using the wrong event there? 

Other platforms have a PlayerLoginEvent or similar which is called when the player instance is first created - does Forge have an equivalent? It looks like `PlayerEvent.LoadFromFile` was chosen originally but I am not sure if that was the correct choice.

Reply:
I dug in deeper and was able to clean this up a bit. The original issue here is that Forge will check command permissions before our user capability was initialised. However, rather than attempting to initialise, I now grab the user directly from LP if we are still in the `pre-login` state. I also changed the `PlayerEvent.LoadFromFile` to `PlayerLoggedInEvent` as it would not always be called on login. Everything seems to work as expected. 

Only question I have now is, should vanilla OP'd users have full access to LP commands? Because currently they do not but I recall other platforms working this way. It does not seem to be related to my changes.